### PR TITLE
APP-475: Fix notification demo flow

### PR DESCRIPTION
**Pull request by RostakaGmfun**
/cc @Slash32 @socallmemaybe @GamovCoder

### CODE REVIEW

**Comment by rasendubi** [comment]
`QA Review fixes` is not a good description. Please, describe what has been found and what this commit does. (Basically, you're rarely interested in how the issue was found.)

**Comment by Slash32** [comment]
@kaa-leeroy test this please

**Comment by Slash32** [comment]
@socallmemaybe @GamovCoder @rasendubi please review.

File: notification/source/cpp/src/KaaDemo.cpp [hunk 1]
```diff
@@ -148,32 +111,67 @@ class TopicListListener : public INotificationTopicListListener {
...
+        } catch(std::exception& e) {
+            std::cout << (boost::format("Incorrect topicId: '%1%'") % input) << std::endl;
+            continue;
+        }
+
+        try {
+            kaaClient->subscribeToTopic(topicId);
+            kaaClient->syncTopicSubscriptions();
+            std::cout << (boost::format("Subscribed to topic '%1%'") % topicId) << std::endl;
+        } catch (KaaException &e) {
```
Comment (ohamov) [inline notification/source/cpp/src/KaaDemo.cpp]: const

File: notification/source/cpp/src/KaaDemo.cpp [hunk 2]
```diff
@@ -148,32 +111,67 @@ class TopicListListener : public INotificationTopicListListener {
...
+        std::unique_lock<std::mutex> lock(updatedMutex);
+        if (!updatedVar.wait_for(lock, std::chrono::seconds(TOPIC_LIST_UPDATE_TIMEOUT),
+                    [&updatedFlag] { return updatedFlag; })) {
+            std::cout << "Timed out waiting for topic list update" << std::endl;
+            kaaClient->stop();
+            return 0;
+        }
+    }
+
+    for (;;) {
```
Comment (ohamov) [inline notification/source/cpp/src/KaaDemo.cpp]: `while (true)`

File: notification/source/cpp/src/KaaDemo.cpp [hunk 3]
```diff
@@ -148,32 +111,67 @@ class TopicListListener : public INotificationTopicListListener {
...
+        std::string input;
+        std::cin >> input;
+
+        if (input == "quit") {
+            break;
+        }
+
+        try {
+            topicId = std::stoll(input);
+        } catch(std::exception& e) {
```
Comment (ohamov) [inline notification/source/cpp/src/KaaDemo.cpp]: cs, const

File: notification/source/cpp/src/KaaDemo.cpp [hunk 4]
```diff
@@ -148,32 +111,67 @@ class TopicListListener : public INotificationTopicListListener {
...
 
     {
-        std::unique_lock<std::mutex> lock(quitMutex);
-        doQuit.wait(lock, [&quitFlag] { return quitFlag; });
+        std::unique_lock<std::mutex> lock(updatedMutex);
+        if (!updatedVar.wait_for(lock, std::chrono::seconds(TOPIC_LIST_UPDATE_TIMEOUT),
+                    [&updatedFlag] { return updatedFlag; })) {
+            std::cout << "Timed out waiting for topic list update" << std::endl;
+            kaaClient->stop();
+            return 0;
```
Comment (ohamov) [inline notification/source/cpp/src/KaaDemo.cpp]: use `EXIT_FAILURE` or `EXIT_SUCCESS` everywhere along the demo.

**Review by ohamov — CHANGES_REQUESTED** [review]

**Comment by am1kk** [comment]
@kaa-leeroy test this please

**Pull request merged**